controller/cmd/tap/main.go

```diff
@@ -33,11 +40,98 @@ func main() {
 	stop := make(chan os.Signal, 1)
 	signal.Notify(stop, os.Interrupt, syscall.SIGTERM)
 
-	server, lis, err := tap.NewServer(*addr, *tapPort, *kubeConfigPath)
+	clientSet, err := k8s.NewClientSet(*kubeConfigPath)
+	if err != nil {
+		log.Fatalf("failed to create Kubernetes client: %s", err)
+	}
+
+	replicaSets, err := k8s.NewReplicaSetStore(clientSet)
+	if err != nil {
+		log.Fatalf("NewReplicaSetStore failed: %s", err)
+	}
+	err = replicaSets.Run()
+	if err != nil {
+		log.Fatalf("replicaSets.Run() failed: %s", err)
+	}
+
+	// index pods by deployment
+	deploymentIndex := func(obj interface{}) ([]string, error) {
+		pod, ok := obj.(*v1.Pod)
+		if !ok {
+			return nil, fmt.Errorf("object is not a Pod")
+		}
+		deployment, err := replicaSets.GetDeploymentForPod(pod)
+		if err != nil {
+			log.Debugf("Cannot get deployment for pod %s: %s", pod.Name, err)
+			return []string{}, nil
+		}
+		return []string{deployment}, nil
+	}
+
+	pods, err := k8s.NewPodIndex(clientSet, deploymentIndex)
```

Comment:
to confirm, all of this was refactored out of `NewServer` and is not changed. and this is required to support the `Tap` call?

Reply:
You are correct this code was moved out of `server.go` into `main.go`. The motivation was to enable testing via dependency injection in `server_test.go`. Now, `NewServer` accepts fully initialized clients, rather than a Kubernetes config file.